Prevent memory leaks on reinstantiation, open up cookies, fix compilation of php >=7.4

seanmorris opened the pull request:
https://github.com/WordPress/playground-tools/pull/110 depends on this PR:

**Allow PHP instances to stop & restart without leaking memory.**

* Expose PHP's exit() function to javascript to allow instances to be shutdown and garbage collected.
* Tweaks load-php-runtime.ts to delete local references to instances upon shutdown.

**Allow calling code to access cookies inside PHPBrowser**

* `PHPBrowser`'s `setCookies` & `serializeCookies` are no longer private methods, allowing multiple instances to share the same login.

**Allow PHP >= 7.4 to compile without failure.**

* ~~Patches were updated for PHP >=7.4 to account for deprecated macros: https://github.com/php/php-src/issues/11678~~
* All versions of PHP >= 8.0 have been updated to the latest minor versions, necessitating some tweaks to the existing patch lines.

**EMSDK Version Locked**

* Emsdk is now locked to use version `3.1.43` rather than `latest`. This makes the dockerfile deterministic, as minor version updates to emscripten are not always 100% 1:1 compatible.
* Recommend not updating that value without running tests first. A dockerfile is a lot like a lockfile. Versions should be committed.

**Dockerfile Re-ordering**

* Moved steps that use PHP_VERSION as far down as possible. All previous steps are identical, so a build of all PHP versions can re-use all these layers from cache. Conversely, if a previous layer uses PHP_VERSION, they will not be considered the same and the cache will not apply. Ordering the build in this way should speed things up a bit as now it can skip to step 54 after running those for the first build.
* OpenSSL no longer copies `man` page documentation as part of the build process, as this is relatively pointless inside a build container.

**Sourcemaps & Debugging**

* The directory structure where the final WASM files live has been shuffled a bit to account for sourcemaps. Sourcemaps may be now be built with `npm run recompile:php:node -- --WITH_SOURCEMAPS=yes`. PHP sources will be copied to each version directory, and sourcemaps can be loaded into chrome devtools via a right click on WASM disassembled code > add source map > `file://[ABSOLUTE LOCAL PATH TO .wasm.map FILE]`

**Iconv**

* libiconv is added to the build: https://github.com/WordPress/wordpress-playground/issues/666

adamziel commented on packages/php-wasm/compile/Dockerfile:
```diff
@@ -950,6 +963,7 @@ RUN source /root/emsdk/emsdk_env.sh && \
     -s ASSERTIONS=0                  \
     -s ERROR_ON_UNDEFINED_SYMBOLS=0  \
     -s NODEJS_CATCH_EXIT=0           \
+	-s NODEJS_CATCH_REJECTION=0      \
```
What are the consequences of this?

seanmorris replied:
@adamziel 

By default, with the current code, any and all unhandled promise rejections will be sent to a handler registered by emscripten. This handler will then simply throw the error as an exception. 

Unhandled rejections and uncaught exceptions are treated similarly from Node 15 onward, so we don't need to explicitly throw an exception when a promise rejection is unhandled.

The problem is that these (relatively inert) handlers are never cleaned up, so when we create and destroy many php instances in a single process, they stack up. Node JS will throw a warning once there are 10 handlers on this event, and then will not register any more handlers, since it's a suspected memory leak.

Setting that switch to `0` prevents this handler from being added in the first place.

This is chunk of code that is being removed from the emscripten-generated javascript:

https://github.com/WordPress/wordpress-playground/blob/90d6042592cd9037158792c4066db81367a8449b/packages/php-wasm/node/public/php_7_1.js#L116-L118

adamziel replied:
You're right! Sweet:

> As of Node.js 15, the default mode for unhandledRejection is changed to throw (from warn). In throw mode, if an unhandledRejection hook is not set, the unhandledRejection is raised as an uncaught exception

https://nodejs.org/en/blog/release/v15.0.0#throw-on-unhandled-rejections---33021

adamziel replied:
Here's the relevant code from Emscripten repo to confirm this is indeed the only effect the flag has:

https://github.com/emscripten-core/emscripten/blob/cbf4256d651455abc7b3332f1943d3df0698e990/src/shell.js#L248-L258

seanmorris replied:
> Here's the relevant code from Emscripten repo to confirm this is indeed the only effect the flag has:
> 
> https://github.com/emscripten-core/emscripten/blob/cbf4256d651455abc7b3332f1943d3df0698e990/src/shell.js#L248-L258

Drives me insane that they didn't include cleanup logic by default...

adamziel commented on packages/php-wasm/compile/Dockerfile:
```diff
@@ -1036,16 +1050,16 @@ RUN \
         # wrapper function.
         /root/replace.sh $'s/ENVIRONMENT_IS_([A-Z]+)\s*=\s*(true|false)/ENVIRONMENT_IS_$1=RuntimeName==="$1"/g' /root/output/php.js && \
-        /root/replace.sh 's/var ENV\s*=\s*\{\}/var ENV = PHPLoader.ENV || {}/g' /root/output/php.js && \
+        /root/replace.sh 's/var ENV\s*=\s*\{\}/var ENV = PHPLoader.ENV || {}/g' /root/output/php.js; \
```
Why replace `&&` with `;` here? For context, `replace.sh` is designed to return an error code when the pattern string can't be found.

seanmorris replied:
@adamziel 

That may be an artifact of some testing I was doing. However, going forward, what would you think of this? We could use `SHELL ["/bin/bash", "-euc", -"o pipefail"]` at the start of the file. This would cause the process to immediately stop and report on ANY error or unset variable.

If we REALLY needed to accept a non-zero exit code, we could use the notation `command || true` to ensure the shell always receives an exit code of 0.

adamziel replied:
@seanmorris Lovely idea! It should work nicely as a follow-up PR as I worry we may need to add quite a few occurrences of `|| true` which could spin into a longer discussion about the benefits, coding styles, alternatives etc. Let's just stick  with `&&` for now unless there's a specific reason to switch to `;`.

seanmorris replied:
@adamziel We could start to implement this by adding `set -eu -o pipefail` as the first line to individual `RUN` commands. In this case the effect will only apply to that particular `RUN`, and once we've got good saturation on that front, migrating to `SHELL ["/bin/bash", "-euc", -"o pipefail"]` will be trivial.

Edit: I actually prefer `-eux` to `-eu` since it will print each command as it runs, instead of just as one large chunk as the `RUN` begins. Let me know your thoughts on that!

seanmorris replied:
@adamziel I noticed that the `if` statements are suffixed with `;`'s instead of `&&`, i.e. `fi; \` rather than `fi && \`, is this expected? I think some failures could be missed in this case, so I added the `set` command from above to the start of the section.

https://github.com/WordPress/wordpress-playground/blob/93bf38f47be4ec5ed08b003e145c4df37eb799c8/packages/php-wasm/compile/Dockerfile#L1061

adamziel replied:
Sounds good! Let's treat these two RUN commands like this in this PR, and work out the general Dockerfile style later on.

> I noticed that the if statements are suffixed with ;'s instead of &&, i.e. fi; \ rather than fi && \, is this expected?

That's a bug. I vaguely remember Docker or bash insisted that using `&&` was a syntax error and I just went for `;` at that time. I'm glad we're revisiting this.

adamziel commented on packages/php-wasm/compile/Dockerfile:
```diff
@@ -38,8 +38,8 @@ RUN set -euxo pipefail;\
 RUN git clone https://github.com/emscripten-core/emsdk.git && \
-    ./emsdk/emsdk install latest && \
-    /root/emsdk/emsdk activate latest
+    ./emsdk/emsdk install 3.1.43 && \
```
good call!

seanmorris replied:
Thank you!

adamziel commented on packages/php-wasm/compile/build-assets/php7.4.patch:
```diff
@@ -34,5 +34,34 @@ diff --git a/php-src/ext/standard/file.c b/php-src/ext/standard/file.c
+ }
+
+-# ifdef COOKIE_SEEKER_USES_OFF64_T
+-static int stream_cookie_seeker(void *cookie, off64_t *position, int whence)
++static int stream_cookie_seeker(void *cookie, off_t *position, int whence)
```
Nice! I know we need these patches to suppress emscripten errors, but I don't have a good understanding why. @seanmorris do you have any insights as to what might be root cause? Is it just that WebAssembly is 32bit and 64bit data types don't fly by default?

seanmorris replied:
@adamziel 
Actually it seems that under the hood, musl was using some macro-magic to map those to the standard names anyhow,  i.e. `off64_t` -> `off_t`. Musl is now deprecating those macros, so we need to account for that one way or another.

https://musl.libc.org/releases.html

> On the API level, the legacy "LFS64" ("large file support") interfaces, which were provided by macros remapping them to their standard names (#define stat64 stat and similar) have been deprecated and are no longer provided under the _GNU_SOURCE feature profile, only under explicit _LARGEFILE64_SOURCE. The latter will also be removed in a future version. Builds broken by this change can be fixed short-term by adding -D_LARGEFILE64_SOURCE to CFLAGS, but should be fixed to use the standard interfaces.

_**See also**_: https://github.com/php/php-src/issues/11678

adamziel replied:
It sounds like we could provide our own mapping for the standard names instead of patching the code to use different names? 🤞

seanmorris replied:
@adamziel Yea that's completely possible. Am I correct in assuming you're suggesting we need to create macros to cover what we lost in the deprecation?

adamziel replied:
@seanmorris yes! Perhaps even more than that. In my ideal world, we wouldn't need to patch types at all. For example, `php7.1.patch` changes `zend_ulong` to `size_t`:

```
-static int pdo_sqlite_stmt_get_col(pdo_stmt_t *stmt, int colno, char **ptr, zend_ulong *len, int *caller_frees)
+static int pdo_sqlite_stmt_get_col(pdo_stmt_t *stmt, int colno, char **ptr, size_t *len, int *caller_frees)
```

Now, this one sounds like a separate problem whatsoever and may be entirely out of scope here, but perhaps it would just take one more macro to deal with that one as well?

seanmorris replied:
@adamziel I'd need to look into the reasoning behind that change before I can comment on it, but lets go ahead and open a github issue to track this, it doesn't seem like a big lift to alias a couple of names if that's all that needs to be done.

adamziel replied:
@seanmorris I opened a separate issue to track this: https://github.com/WordPress/wordpress-playground/issues/652

adamziel commented:
@seanmorris Impressive PR, yay and thank you!  🎉  I think the overall implementation is solid. Still, I have a few questions.

> Expose PHP's exit() function to javascript to allow instances to be shutdown and garbage collected

Are they actually getting garbage collected if you check with memory profiler? That would be amazing! I've been toying with that for a while and couldn't get the JavaScript VM to actually free up the memory. Some of was related to [how the PHP instance may remain referenced in stack trace](https://github.com/WordPress/wordpress-playground/issues/450), but I also remember I couldn't quite destroy all the HEAP instances.

adamziel commented on packages/php-wasm/universal/src/lib/php-browser.ts:
```diff
@@ -111,7 +111,8 @@ export class PHPBrowser implements RequestHandler {
 	get documentRoot() {
 		return this.requestHandler.documentRoot;
 	}
-	#setCookies(cookies: string[]) {
```
I don't have a strong opinion here, but I'm curious what problem does this solve?

seanmorris replied:
That allows us to share cookies between multiple PHP instances. Without that code, we'll get some 403s during handoff between instances due to mismatched authentication tokens.

https://github.com/WordPress/playground-tools/pull/110/files#diff-eeed5d391a89564a5ff176b28d0ab5a31e01b94fb75f4ccc1d67f80297586595R194-R199

adamziel replied:
Ah gotcha. Thank you for explaining!

seanmorris commented:
> @seanmorris Impressive PR, yay and thank you! 🎉 I think the overall implementation is solid. Still, I have a few questions.
> 
> > Expose PHP's exit() function to javascript to allow instances to be shutdown and garbage collected
> 
> Are they actually getting garbage collected if you check with memory profiler? That would be amazing! I've been toying with that for a while and couldn't get the JavaScript VM to actually free up the memory. Some of was related to [how the PHP instance may remain referenced in stack trace](https://github.com/WordPress/wordpress-playground/issues/450), but I also remember I couldn't quite destroy all the HEAP instances.

I did check the memory profiler & used the heap inspector to find dangling references, which is why I modified `packages/php-wasm/universal/src/lib/load-php-runtime.ts` to patch the exit function of the instances it handles on the fly.  I'm **very** happy to report that memory usage stays stable while instances are jockeyed in and out, and everything is cleaned up nicely.

https://github.com/WordPress/wordpress-playground/pull/639/files#diff-bad528da156c5e670aa2d62bf641134058dabcb126aaa7bb4c2ff3c13f706984R172-R183

Edit: Originally, that code was keeping any and all instances it created in an array, I've modified it to use a `Map` object, which is a little easier to deal with. The keys of later objects won't change as earlier keys are removed, nor will we end up with a "sparse" array that's full of holes. This makes it much easier to clean up afterward.

adamziel commented:
@seanmorris this is amazing! Would you mind posting step-by-step instructions on how to check the memory profile alongside the profile file and perhaps a video? I'm sure we'll re-introduce the memory leak over and over again and will be returning to this PR looking for anything helpful. Anything useful we can leave for our future selves will save a ton of time.

Also, thinking out loud, what would a regression test look like here? It would be lovely to have CI constantly verifying the memory leak isn't actually getting reintroduced.

seanmorris commented:
> @seanmorris this is amazing! Would you mind posting step-by-step instructions on how to check the memory profile alongside the profile file and perhaps a video? I'm sure we'll re-introduce the memory leak over and over again and will be returning to this PR looking for anything helpful. Anything useful we can leave for our future selves will save a ton of time.
> 
> Also, thinking out loud, what would a regression test look like here? It would be lovely to have CI constantly verifying the memory leak isn't actually getting reintroduced.

The memory usage of the process with this bug will grow VERY quickly if the memory leaks in this fashion. To ensure the leak is fixed, we'd need to create and destroy many instances in rapid succession, **ensure garbage collection takes place**, then check the process memory usage. If it stays stable, then we're in the clear.

adamziel commented:
Sounds good @seanmorris! Would you please add a test like that to this PR?

seanmorris commented:
> Sounds good @seanmorris! Would you please add a test like that to this PR?

I can do that.

seanmorris commented on packages/php-wasm/compile/build-assets/php_wasm.c:
```diff
@@ -4,6 +4,8 @@
  * This file abstracts the entire PHP API with the minimal set
  * of functions required to run PHP code from JavaScript.
  */
+#define off64_t off_t
```
@adamziel This one line replaces those off64_t patches.

adamziel replied:
Hm `libphp.a` is built in a separate step before `php_wasm.c` is built and linked. Does defining the macro here actually do the trick?

seanmorris replied:
Let me dig a little deeper, although I do believe ~~symbols are preserved in .a files.~~

Symbol preservation is irrelevant.

seanmorris replied:
@adamziel This was driving me insane but I finally realized what was going on here.

musl-**1.2.4** is the version that drops the LFS64 macros:
https://musl.libc.org/releases.html

We're now using emsdk 3.1.43, and version 3.1.4**4** is the version that updates to musl-1.2.4:
https://github.com/emscripten-core/emscripten/blob/main/ChangeLog.md#3144---072523

So the current emsdk is locked to a version without that problem. This will prevent "bit-rot" on the repository and allow this to be prioritized and addressed as bandwidth allows, rather than as an emergency/blocker for another issue. Until we need to update, this is a non-issue.

(By "bit-rot", I mean unexpected changes to behavior with no changes to the codebase.)

adamziel commented on packages/php-wasm/node/src/test/php.spec.ts:
```diff
@@ -759,5 +759,45 @@ describe.each(['7.0', '7.1', '7.3', '7.4', '8.0', '8.1'])(
+		it('Does not leak memory when creating and destroying instances', async () => {
+			let refCount = 0;
+
+			const registry = new FinalizationRegistry(() => --refCount);
```
TIL about FinalizationRegistry, smart!

adamziel commented on packages/php-wasm/node/src/test/php.spec.ts:
```diff
@@ -759,5 +759,45 @@ describe.each(['7.0', '7.1', '7.3', '7.4', '8.0', '8.1'])(
 			}
 		});
+
+		it('Does not leak memory when creating and destroying instances', async () => {
```
Lovely, thank you!

adamziel commented on packages/php-wasm/compile/Dockerfile:
```diff
@@ -1055,5 +1069,13 @@ RUN \
-        mv /root/output/php.wasm "/root/output/$WASM_FILENAME"
\ No newline at end of file
+        mv /root/output/php-module.js "/root/output/$JS_FILENAME"; \
+        # mv /root/output/php.wasm "/root/output/$WASM_FILENAME";
```
Seems like we can remove this line?

```suggestion
```

adamziel commented on packages/php-wasm/compile/Dockerfile:
```diff
@@ -1055,5 +1069,13 @@ RUN \
\ No newline at end of file
+        mv /root/output/php-module.js "/root/output/$JS_FILENAME"; \
+        # mv /root/output/php.wasm "/root/output/$WASM_FILENAME";
+        mkdir -p /root/output/${PHP_VERSION_ESCAPED}/; \
```
Let's stick with `&&` in this and next lines until the Dockerfile uses pipefail by default

seanmorris replied:
Same idea: https://github.com/WordPress/wordpress-playground/pull/639#discussion_r1346673235

adamziel commented on packages/php-wasm/compile/build-assets/php8.0.patch:
```diff
@@ -1,7 +1,8 @@
 --- a/php-src/ext/standard/file.c
 +++ b/php-src/ext/standard/file.c
-@@ -1691,18 +1691,32 @@ PHPAPI int php_copy_file_ctx(const char *src, const char *dest, int src_flg, php
+@@ -1625,17 +1625,31 @@ PHPAPI int php_copy_file_ctx(const char *src, const char *dest, int src_flg, php
```
Should this one also restore the full `php_stream_context` definition and the closing `)`?

```suggestion
@@ -1625,17 +1625,31 @@ PHPAPI int php_copy_file_ctx(const char *src, const char *dest, int src_flg, php_stream_context *ctx)
```

seanmorris replied:
> Should this one also restore the full `php_stream_context` definition and the closing `)`?

https://en.wikipedia.org/wiki/Diff#Unified_format

Everything after the second `@@` on that line is essentially a comment. That's just the header line for the chunk of the diff, it doesn't contain any replacement information. The important bit is the location data between the `@@`'s.

adamziel commented on packages/php-wasm/compile/build-assets/php8.1.patch:
```diff
@@ -4,24 +4,26 @@ diff --git a/php-src/sapi/embed/php_embed.c b/php-src/sapi/embed/php_embed.c
  	php_stream *srcstream = NULL, *deststream = NULL;
  	int ret = FAILURE;
  	php_stream_statbuf src_s, dest_s;
+ 	int src_stat_flags = (src_flg & STREAM_DISABLE_OPEN_BASEDIR) ? PHP_STREAM_URL_STAT_IGNORE_OPEN_BASEDIR : 0;
```
Great catch!

adamziel commented on .github/workflows/ci.yml:
```diff
@@ -27,7 +27,7 @@ jobs:
             - uses: actions/checkout@v3
             - uses: ./.github/actions/prepare-playground
-            - run: npx nx affected --target=test --configuration=ci
+            - run: node --expose-gc node_modules/nx/bin/nx affected --target=test --configuration=ci
```
If this is needed for the memory leak test to pass then local tests should run with this flag, too. Otherwise that one test will always fail. Perhaps updating the `test` script in `package.json` would be a good start.

It would be also nice to assert that `--expose-gc` was used and provide a useful error message explaining how you need to run `node --expose-gc  node_modules/nx/bin/nx` and not just `nx`.

seanmorris replied:
@adamziel This is what I wanted to bring up in the meeting. Slipped my mind. Is there a way in `nx` to include this switch by default?

I think that would be better than creating a special case with a custom invocation here. I can make the change you mentioned, but let me know what you think.

adamziel replied:
@seanmorris not that I'm aware of. The [vitest executor](https://nx.dev/nx-api/vite/executors/test) doesn't list any node-specific options, neither does the [vitest](https://vitest.dev/config/) library itself. Running `nx` seems to be forking into a `nx` daemon process and an executor process where all node `execArgv` options are passed to the executor process. Other than that, I don't see a way to specify `execArgv` for the executor process in the general case. I guess that's something `nx` itself would need to support.

One thing we could do is define the `test` command using the `nx:run-commands` executor, like this:

```json
		"publish": {
			"executor": "nx:run-commands",
			"options": {
				"command": "node tools/scripts/publish.mjs php-wasm-node {args.ver} {args.tag}",
				"parallel": false
			},
			"dependsOn": ["build"]
		}
```

It could enforce `--expose-gc` but would just require some fiddling to find the right `vitest` command to run.

seanmorris replied:
@adamziel Hmmm, perhaps we should keep things working the way we've got it right now then, and revisit if we end up having issues with it.

I don't think that command is going to be called manually much.

seanmorris replied:
@adamziel  https://github.com/WordPress/wordpress-playground/pull/639/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R38

adamziel commented:
It seems this PR is almost there, I just left a couple more notes.

adamziel commented on packages/php-wasm/node/src/test/php.spec.ts:
```diff
@@ -759,5 +759,55 @@ describe.each(['7.0', '7.1', '7.3', '7.4', '8.0', '8.1'])(
+
+		it('Does not leak memory when creating and destroying instances', async () => {
+			if (!global.gc) {
+				console.error(
```
👍  nice!

adamziel commented:
This PR ships PHP for node but also affects PHP for the web. Let's add that, too. Ideally GitHub actions would handle that but we're not there yet.

adamziel commented:
I just ran into the following error when building the web version with source maps:

```
$ npm run recompile:php:web:7.4 -- --WITH_SOURCEMAPS=yes
...
#94 0.194 sed: can't read /root/output/7_4_33/php.wasm.map: No such file or directory
```

adamziel commented:
I thought it's a typo in a path but the map file seems to be actually missing

seanmorris commented:
> I thought it's a typo in a path but the map file seems to be actually missing

@adamziel I originally had it set to only build the sourcemaps for `node`, not `web`. Fixed:

https://github.com/WordPress/wordpress-playground/pull/639/files#diff-254451283e52699f8082ac4dbfc2e589784be646e560bfb3a8e6b499eb8527adR556-R564

seanmorris commented:
> This PR ships PHP for node but also affects PHP for the web. Let's add that, too. Ideally GitHub actions would handle that but we're not there yet.

Are you suggesting I just rebuild PHP for `web` like I did for `node`?

adamziel commented:
Yup :) I see web files are rebuilt now, thanks. I’ll test it.

adamziel commented on packages/php-wasm/compile/Dockerfile:
```diff
@@ -1055,5 +1101,12 @@ RUN \
+        mkdir -p /root/output/${PHP_VERSION_ESCAPED}/; \
+		mv /root/output/php.wasm "/root/output/${PHP_VERSION_ESCAPED}/$WASM_FILENAME"; \
+		if [ "${WITH_SOURCEMAPS}" = "yes" ]; then \
+			sed -i 's/..\/php-src/php-src/g' /root/output/php.wasm.map; \
```
Let's remove the `.git` directory:

```suggestion
			rm -rf /root/php-src/.git; \
			sed -i 's/..\/php-src/php-src/g' /root/output/php.wasm.map; \
```

Preserving it triggers the following error on my end:

```
Running docker run --name php-wasm-tmp --rm -v packages/php-wasm/web/public:/output php-wasm sh -c cp -rf /root/output/* /output && mkdir -p /output/terminfo/x  ...
cp: cannot create regular file '/output/7_4_33/php-src/.git/objects/pack/pack-7063960767cc0b8bc15af3550492d4c4145a5efa.idx': Permission denied
```

adamziel replied:
Unescaped `.` stands for any character, so `../../root/php-src` gets replaced with `../../rophp-src`. 

The following regexp does the trick:

```suggestion
			# Make the source map paths relative to the .wasm file, not the build directory
			sed -i 's#"\.\./\.\./root/#"#g' /root/output/php.wasm.map
```

We could [avoid this replacement entirely](https://github.com/emscripten-core/emscripten/blob/e18753e6fbbf2d690e5a750165c693bb6bfec122/tools/building.py#L1162) by building to `/root`, but that would require more changes to the pipeline so let's leave it for now.

adamziel commented on packages/php-wasm/compile/Dockerfile:
```diff
@@ -1055,5 +1101,12 @@ RUN \
+		if [ "${WITH_SOURCEMAPS}" = "yes" ]; then \
+			sed -i 's/..\/php-src/php-src/g' /root/output/php.wasm.map; \
+			mv /root/output/php.wasm.map "/root/output/${PHP_VERSION_ESCAPED}/php.wasm.map"; \
+			cp -r /root/php-src "/root/output/${PHP_VERSION_ESCAPED}/php-src"; \
```
Let's remove the `.git` directory:

```suggestion
			rm -rf /root/php-src/.git; \
			cp -r /root/php-src "/root/output/${PHP_VERSION_ESCAPED}/php-src"; \
```

Preserving it triggers the following error on my end:

```
Running docker run --name php-wasm-tmp --rm -v packages/php-wasm/web/public:/output php-wasm sh -c cp -rf /root/output/* /output && mkdir -p /output/terminfo/x  ...
cp: cannot create regular file '/output/7_4_33/php-src/.git/objects/pack/pack-7063960767cc0b8bc15af3550492d4c4145a5efa.idx': Permission denied
```

adamziel commented:
I found the following issues with source maps on the web:

* php.js files produced when the source maps are enabled do not contain the `import dependencyFilename from './7_4_33/php_7_4.wasm'; ` line and start with `export { dependencyFilename }; ` which results in a JS error
* The source map doesn't load

<img width="867" alt="CleanShot 2023-10-10 at 11 03 06@2x" src="https://github.com/WordPress/wordpress-playground/assets/205419/af487003-4b2e-49c9-9969-e243b705a085">

### Import

the `import` is missing due to the following error that's only visible after changing the command separators from `&&` to `;`:

```
No matches found for s/var ENV\s*=\s*\{\}/var ENV = PHPLoader.ENV || {}/g in file /root/output/php.js
```

So the good news is that pipefail actually lets us know what's wrong. Now, how to fix this? The unminified JS module defines ENV as follows:

```
 var ENV = {
  };
```

Notice the newline between `{` and `}`. Adjusting the regexp to `s/var ENV\s*=\s*\{/var ENV = PHPLoader.ENV || {/g` does the trick.

### Source map loading

We need to use the `--source-map-base http://host:port/base-path-to-source-maps-and-sources/` option for `emcc` so the source map is loaded via `http://` and not `wasm://`. Also, there's [an issue with path normalization](https://github.com/WordPress/wordpress-playground/pull/639#discussion_r1352068100) in Dockerfile.

---

I played around, left some comments, and got it to work with a few more tweaks. @seanmorris Feel free to resolve all the partial comments I left and cherry-pick https://github.com/WordPress/wordpress-playground/commit/0c2beda9 which comes with all my changes.

<img width="807" alt="CleanShot 2023-10-10 at 11 42 21@2x" src="https://github.com/WordPress/wordpress-playground/assets/205419/3cb36d3f-84df-488f-8295-0556c532536e">

adamziel commented:
Actually, spending some more time here, the `import` is missing due to the following error that's only visible after changing the command separators from `&&` to `;`:

```
No matches found for s/var ENV\s*=\s*\{\}/var ENV = PHPLoader.ENV || {}/g in file /root/output/php.js
```

So the good news is that pipefail actually lets us know what's wrong. Now, how to fix this? The unminified JS module defines ENV as follows:

```
 var ENV = {
  };
```

Notice the newline between `{` and `}`. Adjusting the regexp to `s/var ENV\s*=\s*\{/var ENV = PHPLoader.ENV || {` does the trick.

adamziel commented:
Okay I got it to work:

<img width="807" alt="CleanShot 2023-10-10 at 11 42 21@2x" src="https://github.com/WordPress/wordpress-playground/assets/205419/3cb36d3f-84df-488f-8295-0556c532536e">

adamziel commented:
One last thing that bothers me is that the size of PHP each `.wasm` file increased by ~0.6MB on the web. Can we keep it small?

I tested the following hypotheses and this increase is **not** caused by:

* Pinning emscripten to `3.1.43` – doing this on trunk didn't increase the binary size
* Iconv – Removing it entirely from the Dockerfile did not decrease the binary size
* Source maps – Removing all `if` referring to `WITH_SOURCEMAPS` did not decrease the binary size
* `NODEJS_CATCH_REJECTION`
* Exporting the `_exit` function

Found it on my 6th attempt. This line is now used on the web and not just in node.js:

```
echo -n ' -g2 ' >> /root/.emcc-php-wasm-flags; \
```

It preserves the symbol names. Let's not do that in the production build.

adamziel commented:
@seanmorris almost there! Here's what we need in order to merge:

* Rebase
* Please review https://github.com/WordPress/wordpress-playground/commit/0c2beda9 and apply it here if it looks good to you
* Confirm source maps still work on Node.js with the above patch – I only tried my changes on the web platform
* Rebuild PHP binaries for the web to avoid shipping larger–than–necessary assets on playground.wordpress.net

adamziel commented on packages/php-wasm/compile/Dockerfile:
```diff
@@ -578,13 +577,19 @@ RUN if [[ "${PHP_VERSION:0:1}" -le "7" && "${PHP_VERSION:2:1}" -le "3" ]]; then
+	if [ "${WITH_SOURCEMAPS}" = "yes" ]; then \
+		echo -n ' -g3 -gsource-map' >> /root/.emcc-php-wasm-flags; \
+	else \
+		echo -n ' -g2 ' >> /root/.emcc-php-wasm-flags; \
```
Let's only preserve the symbols (`-g2`, `-g3`) in Node.js and not on the web because that adds like 600kb to the bundle size. Well, unless we're doing sourece maps for the web. So Node.js + Web with source maps.

seanmorris replied:
https://github.com/WordPress/wordpress-playground/pull/639/files#diff-254451283e52699f8082ac4dbfc2e589784be646e560bfb3a8e6b499eb8527adR588

dmsnell commented on packages/playground/blueprints/src/lib/steps/run-sql-file.ts:
```diff
@@ -0,0 +1,49 @@
+	const runPhp = await playground.run({
+		code: `<?php
+	require_once '/wordpress/wp-load.php';
+	require_once '/wordpress//wp-content/plugins/Collector/Collector_Restore.php';
```
nit: there's an extra `/` in here

dmsnell commented on packages/playground/blueprints/src/lib/steps/run-sql-file.ts:
```diff
@@ -0,0 +1,49 @@
+ * <code>
+ * {
+ * 	    "step": "runSqlFile",
+ * 		"path": "wordpress.org/plugins"
```
is this path correct? it seems surprising

seanmorris replied:
@dmsnell This was an untracked file that should have been removed.

dmsnell commented on packages/playground/blueprints/src/lib/steps/run-sql-file.ts:
```diff
@@ -0,0 +1,49 @@
+	 */
+	step: 'runSqlFile';
+	/**
+	 * The plugin zip file to install.
```
this comment is wrong I think.

dmsnell commented on packages/playground/blueprints/src/lib/steps/common.ts:
```diff
@@ -25,6 +25,12 @@ export async function fileToUint8Array(file: File) {
+	return new TextDecoder('utf-8').decode(
+		new Uint8Array(await file.arrayBuffer())
+	);
+}
```
should we indicate in the name that this assumes UTF-8? it won't work in the generic sense for files, evident for any image or even many legacy files encoded in `latin1` or other encodings.

in these cases we'll probably get a string full of � symbols because the bytes couldn't be decoded.

seanmorris replied:
@dmsnell I wholeheartedly agree, but this ended up getting refactored out and is unused. I'm removing it now.

dmsnell replied:
even better

adamziel commented on packages/php-wasm/compile/Dockerfile:
```diff
@@ -1061,85 +1082,96 @@ RUN ls /root/output/
-        # We need to patch the syscall to support the MSG_PEEK flag.
-        if [ "$WITH_WS_NETWORKING_PROXY" = "yes" ]; then \
-            /root/replace.sh 's/sock\.sock_ops\.recvmsg\(sock,\s*len\);/sock.sock_ops.recvmsg(sock, len, typeof flags !== "undefined" ? flags : 0);/g' /root/output/php.js && \
-            /root/replace.sh 's/recvmsg:\s*function\(sock,\s*length\)/recvmsg: function(sock, length, flags)/g' /root/output/php.js && \
```
was that pattern wrong?

adamziel replied:
yes it was, good catch!

adamziel commented on packages/php-wasm/universal/src/lib/base-php.ts:
```diff
@@ -526,7 +526,8 @@ export abstract class BasePHP implements IsomorphicLocalPHP {
 					[],
-					[]
+					[],
+					{ async: true }
```
TIL about this option, good spot

adamziel reviewed (approved):
🎉  LET'S SHIP!
